src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs

```diff
@@ -0,0 +1,301 @@
+                Marshal.WriteIntPtr(cxtInt.ClassFactoryDest, nativeIUnknown);
+            }
+            catch (Exception e)
+            {
+                return e.HResult;
+            }
+
+            return 0;
+        }
+
+        private static bool IsLoggingEnabled()
+        {
+#if DEBUG
+            return true;
+#else
+            return false;
+#endif
+        }
+
+        private static void Log(string fmt, params object[] args)
+        {
+            // [TODO] Consider using FrameworkEventSource in release builds
+
+            Debug.WriteLine(fmt, args);
+         }
+
+        private static (Assembly assembly, Type type) FindClassAssemblyAndType(Guid clsid, string[] potentialAssembies)
+        {
+            // Determine what assembly the class is in
+            foreach (string assemPath in potentialAssembies)
+            {
+                Assembly assem;
+                string assemPathLocal = assemPath;
+
+                try
+                {
+                    string extMaybe = Path.GetExtension(assemPath);
+                    if (".manifest".Equals(extMaybe, StringComparison.OrdinalIgnoreCase))
+                    {
+                        assemPathLocal = Path.ChangeExtension(assemPath, ".dll");
+                    }
+
+                    assem = Assembly.LoadFrom(assemPathLocal);
+                }
+                catch (Exception e)
+                {
+                    if (IsLoggingEnabled())
+                    {
+                        Log($"COM Activation of {clsid} failed to load assembly {assemPathLocal}: {e}");
+                    }
+
+                    continue;
+                }
+
+                // Check the loaded assembly for a class with the desired ID
+                foreach (Type t in assem.GetTypes())
+                {
+                    if (t.GUID == clsid)
+                    {
+                        return (assem, t);
+                    }
+                }
+            }
+
+            // [TODO] Check Registry for registration
+
+            const int CLASS_E_CLASSNOTAVAILABLE = unchecked((int)0x80040111);
+            throw new COMException(string.Empty, CLASS_E_CLASSNOTAVAILABLE);
+        }
+
+        private static string[] CreateAssemblyArray(int assemblyCount, IntPtr assemblyList)
+        {
+            var assemblies = new string[assemblyCount];
+
+            unsafe
+            {
+                var spanOfPtrs = new Span<IntPtr>(assemblyList.ToPointer(), assemblyCount);
+                for (int i = 0; i < assemblyCount; ++i)
+                {
+                    assemblies[i] = Marshal.PtrToStringUni(spanOfPtrs[i]);
+                }
+            }
+
+            return assemblies;
+        }
+
+        [ComVisible(true)]
+        internal class BasicClassFactory : IClassFactory2
+        {
+            private readonly Guid classId;
+            private readonly Type classType;
+            private readonly Assembly classAssembly;
```

Comment:
There's no need to store the assembly here - if nothing else it can be asked on the type, but the class should not need to know the assembly to do anything anyway.

Reply:
Done